Adds a how to guide to create coordinate object with custom observer (#7769)
sunpy/sunpy: docs/how_to/create_custom_coords.rst

.. _sunpy-how-to-create-custom-coordinate-objects:

**********************************************
Create Coordinate Objects with Custom Observer
**********************************************

> It is possible to specify an observer other than Earth by providing a `~astropy.coordinates.SkyCoord` object representing the observer's location. This can be useful when working with data from instruments located at specific points in space, such as spacecraft.

Reviewer: ```suggestion
It is possible to specify an observer other than Earth by providing a `~astropy.coordinates.SkyCoord` object representing the observer's location.
This can be useful when working with data from instruments located at specific points in space, such as spacecraft.
```

.. code-block:: python

    >>> from astropy.coordinates import SkyCoord, CartesianRepresentation
    >>> from sunpy.coordinates import frames
    >>> import astropy.units as u

Define the observer's position in Heliocentric coordinates

.. code-block:: python

    >>> observer_coord = SkyCoord(-100*u.arcsec, 500*u.arcsec, frame=frames.Helioprojective)
    >>> observer_coord= SkyCoord(x=-72241.0*u.km, y=361206.1*u.km, z=589951.4*u.km, frame=frames.Heliocentric)
    >>> observer_coord = SkyCoord(70*u.deg, -30*u.deg, frame=frames.HeliographicStonyhurst)
    >>> observer_coord
    <SkyCoord (HeliographicStonyhurst: obstime=None, rsun=695700.0 km): (lon, lat) in deg
        (70., -30.)>

Create a coordinate in Helioprojective with the custom observer

.. code-block:: python

    >>> coord = SkyCoord(100*u.arcsec, 200*u.arcsec, frame=frames.Helioprojective, observer=observer_coord)
    >>> coord
    <SkyCoord (Helioprojective: obstime=None, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=None, rsun=695700.0 km): (lon, lat, radius) in (deg, deg, km)
        (70., -30., 695700.)>): (Tx, Ty) in arcsec
        (100., 200.)>

> In this example, we define the observer's position at -100*u.arcsec along the x-axis and 500*u.arcsec along y-axis in Heliocentric coordinates. This observer coordinate is then used to create a `~astropy.coordinates.SkyCoord` object in the ``Helioprojective`` frame.

Reviewer: ```suggestion
In this example, we define the observer's position at ``-100*u.arcsec`` along the x-axis and ``500*u.arcsec`` along y-axis in Heliocentric coordinates.
This observer coordinate is then used to create a `~astropy.coordinates.SkyCoord` object in the ``Helioprojective`` frame.
```

> .. _sunpy-how-to-create-custom-coordinate-objects:

Reviewer 2: ```suggestion
.. _sunpy-how-to-observer-by-coordinate:
```

> **********************************************
> Create Coordinate Objects with Custom Observer
> **********************************************

Reviewer 2: ```suggestion
************************************
Specifying an Observer by Coordinate
************************************
```

> It is possible to specify an observer other than Earth by providing a `~astropy.coordinates.SkyCoord` object representing the observer's location.
> This can be useful when working with data from instruments located at specific points in space, such as spacecraft.

Reviewer 2: ```suggestion
For coordinate frames that include the observer location as part of the definition (e.g., `~sunpy.coordinates.Helioprojective`), there are two main ways to specify the observer location (the ``observer`` frame attribute).
If it makes sense to designate a planetary body as the observer, one can specify the observer to be that body using a string (e.g., ``observer='earth'``), and the observer will be understood to be situated at the center of that body at the time of the coordinate (the ``obstime`` frame attribute).
However, if the precise location on a planetary body is important, or if the observer is elsewhere in the space (e.g., a spacecraft), one can specify that observer location using a `~astropy.coordinates.SkyCoord` object.
```

.. code-block:: python

    >>> from astropy.coordinates import SkyCoord, CartesianRepresentation
    >>> from sunpy.coordinates import frames
    >>> import astropy.units as u

> Define the observer's position in Heliocentric coordinates

Reviewer 2: ```suggestion
The `~astropy.coordinates.SkyCoord` for the observer location can come from a variety of sources.
If you are working with a `~sunpy.map.Map` with an observer location stored in the metadata, you can retrieve the observer location using the `~sunpy.map.GenericMap.observer_coordinate` property.
If the observer is a major spacecraft, you might be able to obtain its location using :func:`~sunpy.coordinates.get_horizons_coord`.
Here, we will manually specify an observer location in Stonyhurst heliographic coordinates (`~sunpy.coordinates.HeliographicStonyhurst`):
```

.. code-block:: python

>     >>> observer_coord = SkyCoord(-100*u.arcsec, 500*u.arcsec, frame=frames.Helioprojective)
>     >>> observer_coord= SkyCoord(x=-72241.0*u.km, y=361206.1*u.km, z=589951.4*u.km, frame=frames.Heliocentric)

Reviewer 2: ```suggestion
```
These two lines are bad examples of specifying an observer location.  First, they are both observer-based coordinate frames, so there is an implicit `observer='earth'`.  Second, they are both on the surface of the Sun, which is not where any realistic observer would be.

>     >>> observer_coord = SkyCoord(70*u.deg, -30*u.deg, frame=frames.HeliographicStonyhurst)
>     >>> observer_coord
>     <SkyCoord (HeliographicStonyhurst: obstime=None, rsun=695700.0 km): (lon, lat) in deg
>         (70., -30.)>

Reviewer 2: ```suggestion
    >>> observer_coord = SkyCoord(70*u.deg, -30*u.deg, 1*u.AU, frame=frames.HeliographicStonyhurst)
    >>> observer_coord
    <SkyCoord (HeliographicStonyhurst: obstime=None, rsun=695700.0 km): (lon, lat, radius) in (deg, deg, AU)
        (70., -30., 1.)>
```
If you don't specify a distance to `HeliographicStonyhurst`, it will get interpreted as being on the surface of the Sun, which again is not a realistic observer location.

Author: Ah, Understood

> Create a coordinate in Helioprojective with the custom observer

Reviewer 2: ```suggestion
Now we can create a coordinate in helioprojective coordinates (`~sunpy.coordinates.Helioprojective`) using the observer location defined above:
```

.. code-block:: python

    >>> coord = SkyCoord(100*u.arcsec, 200*u.arcsec, frame=frames.Helioprojective, observer=observer_coord)
    >>> coord

>     <SkyCoord (Helioprojective: obstime=None, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=None, rsun=695700.0 km): (lon, lat, radius) in (deg, deg, km)
>         (70., -30., 695700.)>): (Tx, Ty) in arcsec

Reviewer 2: ```suggestion
    <SkyCoord (Helioprojective: obstime=None, rsun=695700.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=None, rsun=695700.0 km): (lon, lat, radius) in (deg, deg, AU)
        (70., -30., 1.)>): (Tx, Ty) in arcsec
```

        (100., 200.)>

> 
> In this example, we define the observer's position at ``-100*u.arcsec`` along the x-axis and ``500*u.arcsec`` along y-axis in Heliocentric coordinates.
> This observer coordinate is then used to create a `~astropy.coordinates.SkyCoord` object in the ``Helioprojective`` frame.

Reviewer 2: ```suggestion
```
